Post notes

ErikRauer opened the pull request.

audreylemeur committed:
Create button to Add Notes

Create button that will be linked to the add notes page from the home page. Furthers issue #12.

audreylemeur committed:
Add unit testing for client side Add Notes

Furthers issue #12.

audreylemeur committed:
Create Add Note Component

Furthers issue #12.

ErikRauer committed:
Fix validators and html of addNoteComponent

ErikRauer committed:
Add routing for addNoteComponent

ErikRauer committed:
Fix issues with addNoteComponent

ErikRauer committed:
Add addNote() method to note service

ErikRauer committed:
Add test for addNote() method in NotesService

ErikRauer committed:
Add content to submitForm() method in addNoteComponent

ErikRauer committed:
Add endpoint for adding new notes

ErikRauer committed:
Add method to NoteController() for adding new notes

ErikRauer committed:
Fix endpoint

ErikRauer committed:
Fix navigation after adding a note

ErikRauer committed:
Add test for adding notes on server side

ErikRauer committed:
Add test for invalid note on server side

ErikRauer committed:
Fix routing for adding a new note

Button now correctly routes to the addNoteComponent

ErikRauer committed:
Fix junit tests not working and add a test for body being too long

ErikRauer committed:
Add e2e testing for New Note page

ErikRauer committed:
Remove warning from add-note e2e spec

ErikRauer committed:
Add e2e test for home page

The test checks that the user can navigate to the add note page

ErikRauer committed:
Merge branch 'master' into post_notes

ErikRauer commented:
Also closes #6 #7

helloworld12321 commented on client/e2e/src/add-note.e2e-spec.ts:
```diff
@@ -0,0 +1,41 @@
+import { E2EUtil } from './e2e.util';
+
+
+describe('Add note:', () => {
```
Low-key, it makes me so happy that you all used the same capitalization and punctuation as the other e2e tests 🙃 That's wonderful!

helloworld12321 commented on client/e2e/src/add-note.e2e-spec.ts:
```diff
@@ -0,0 +1,41 @@
+
+    await page.typeInput('bodyField', E2EUtil.randomText(250));
+    expect(element(by.buttonText('ADD NOTE')).isEnabled()).toBe(false);
+  });
```
This is really good.

helloworld12321 commented on client/e2e/src/add-note.e2e-spec.ts:
```diff
@@ -0,0 +1,41 @@
+      body: E2EUtil.randomText(10)
+    };
+
+    await page.addNote(note);
```
I'd recommend clicking the button instead of calling page.addNote directly :) That way you see that the button works!

helloworld12321 replied:
You even have a method to do that in add-note.po.ts already

helloworld12321 replied:
Oops, never mind, `addNote` is a method of `AddNotePage`—I thought you were calling a method of `AddNoteComponent` named `addNote` instead of going through the UI to do it. Pardon me! 🙇🏻‍♂️ My bad!

helloworld12321 commented on client/e2e/src/home.e2e-spec.ts:
```diff
@@ -0,0 +1,26 @@
+    await browser.wait(EC.urlContains('/new'), 10000);
+
+    url = await page.getUrl();
+    expect(url.endsWith('/new')).toBe(true);
```
A little bit redundant—you're already waiting until the browser contains `/new`, so there's no way that this `expect` could *not* be `true`.

But the test will still fail if the url doesn't contain `/new`—it'll just timeout on line 20. That should be fine then ☺️

helloworld12321 commented on client/src/app/add/add-note.component.ts:
```diff
@@ -0,0 +1,62 @@
+
+  // not sure if this name is magical and making it be found or if I'm missing something,
+  // but this is where the red text that shows up (when there is invalid input) comes from
+  add_note_validation_messages = {
```
Would you be so kind as to rename this to `addNoteValidationMessages`? The underscores are a little weird.

(You'll need to change `add-note.component.html` as well.)

helloworld12321 replied:
For the record, the name is not magical. The thing that's making it be found is that you're referencing it from `add-note.component.html`. 🙂

helloworld12321 commented on client/src/app/add/add-note.component.ts:
```diff
@@ -0,0 +1,62 @@
+    body: [
+      {type: 'required', message: 'Body is required'},
+      {type: 'minlength', message: 'Body must be at least 2 characters long'},
+      {type: 'maxlength', message: 'Body cannot be more than 50 characters long'}
```
Really? 50 characters? Even Twitter gives you more than that 🙃

helloworld12321 commented on client/src/app/add/add-note.component.ts:
```diff
@@ -0,0 +1,62 @@
+      body: new FormControl('', Validators.compose([
+        Validators.required,
+        Validators.minLength(2),
+        Validators.maxLength(250),
```
And then down here you say the maximum length is `250`, not `50`.

helloworld12321 replied:
Can we just bump it up to an even thousand? That's about two paragraphs.

helloworld12321 replied:
(You'd need to change it in the e2e tests as well.)

helloworld12321 replied:
(And possibly the Karma tests)

helloworld12321 replied:
Update: Audrey recommends `280` characters because that's the length of a tweet 🙃

helloworld12321 replied:
(And the server)

helloworld12321 replied:
You know, that might be a lot of work. I'm fine with leaving it at 250.

ErikRauer replied:
I'll increase it to 300

helloworld12321 commented on client/src/app/add/add-note.component.spec.ts:
```diff
@@ -0,0 +1,92 @@
+    fixture.detectChanges();
+    addNoteForm = addNoteComponent.addNoteForm;
+    expect(addNoteForm).toBeDefined();
+    expect(addNoteForm.controls).toBeDefined();
```
Move lines 48-49 into the `should create` test

helloworld12321 commented on client/src/app/add/add-note.component.ts:
```diff
@@ -0,0 +1,62 @@
+    this.addNoteForm = this.fb.group({
+      body: new FormControl('', Validators.compose([
+        Validators.required,
+        Validators.minLength(2),
```
Do we need a minimum length at all? I mean, I don't think it's *harming* anything here, but why do we care if the message is one character long?

ErikRauer replied:
I don't see any situation where the user will be posting a note that's 1 character long.

helloworld12321 replied:
I'd like to formally disagree with this decision. However, I'm willing to abide by it.

helloworld12321 commented on client/src/app/add/add-note.component.html:
```diff
@@ -0,0 +1,31 @@
+<div fxLayout="row">
+  <div fxFlex fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10">
+    <form [formGroup]="addNoteForm" (ngSubmit)="submitForm();">
+    <mat-card class="add-note-card">
```
This file is indented suuuper weirdly.

helloworld12321 commented on server/src/main/java/umm3601/notes/NoteController.java:
```diff
@@ -26,4 +27,14 @@ public void getNotes(Context ctx) {
+    .check((note) -> note.body.length() >= 2 && note.body.length() <= 250).get();
+
+    noteCollection.insertOne(newNote);
+    ctx.status(201);
```
The `201: created` status is a nice touch :)

helloworld12321 commented on server/src/test/java/umm3601/note/NoteControllerSpec.java:
```diff
@@ -99,4 +107,59 @@ public void GetAllNotes() throws IOException {
+  @Test
+  public void AddNote() throws IOException {
+
+    String testNewNote = "{\n\t\"body\": \"Test Note\"}";
```
Remove the `\n\t` here—that's not helping readability any

helloworld12321 commented on server/src/test/java/umm3601/note/NoteControllerSpec.java:
```diff
@@ -99,4 +107,59 @@ public void GetAllNotes() throws IOException {
+    String result = ctx.resultString();
+    String id = jsonMapper.readValue(result, ObjectNode.class).get("id").asText();
+    assertNotEquals("", id);
+    System.out.println(id);
```
Remove this print statement

helloworld12321 commented on server/src/test/java/umm3601/note/NoteControllerSpec.java:
```diff
@@ -99,4 +107,59 @@ public void GetAllNotes() throws IOException {
+
+  @Test
+  public void AddNoteWithTooShortBody() throws IOException {
+    String testNewNote = "{\n\t\"body\":x\n}";
```
Similarly, remove the `\n\t` and the `\n`

helloworld12321 replied:
Also, you'll need to quote the `x`.

helloworld12321 commented on server/src/test/java/umm3601/note/NoteControllerSpec.java:
```diff
@@ -99,4 +107,59 @@ public void GetAllNotes() throws IOException {
+  }
+
+  public void AddNoteWithTooLongBody() throws IOException {
+    String testNewNote = "{\n\t\"body\":";
```
Here as well

helloworld12321 reviewed (changes requested):
Okay, there are some things to modify before merging into master:

1. Reformat `client/src/app/add/add-note.component.html`
2. In `client/src/app/add/add-note.component.spec.ts`, move lines 48-49 into the `should create` test
3. In `client/src/app/add/add-note.component.ts` and `client/src/app/add/add-note.component.html`, rename `add_note_validation_messages` to `addNoteValidationMessages`.
4. Remove the minimum length requirement on notes.
5. In `server/src/test/java/umm3601/note/NoteControllerSpec.java`, remove the print statement on line 128, and remove all instances of `\n` and `\t` from JSON strings.
6. In `server/src/test/java/umm3601/note/NoteControllerSpec.java`, make sure that the body has quotes around it in all JSON strings.

(In general, looks really good, though! 😊)

ErikRauer committed:
Refactor according to Joe's suggestions

Change the e2e test to click the button directly rather than do it through the po file. Also rename validation messages for adding a note to be more in line with the naming of everything else in our project.

ErikRauer committed:
Increase max length of notes to 300 characters

ErikRauer committed:
Refactor according to the rest of Joe's suggestions

Fix the indenting being off in the add-note component HTML. Also made the Json in the NoteControllerSpec easier to read.

helloworld12321 reviewed (commented):
Thanks so much! 😊

helloworld12321 commented:
(Excuse me—just trying to give you that green check mark 🙃)

helloworld12321 reviewed (commented):
Everything is good now

helloworld12321 commented:
(Pardon me—GitHub isn't letting me give you the little green check mark even though you deserve it 😢)

(I'll just second Audrey's check-mark 🙃 Check! ✅)
